packages/flutter/lib/src/gestures/hit_test.dart

```diff
@@ -148,26 +220,60 @@ class HitTestResult {
       'matrix through PointerEvent.removePerspectiveTransform? '
       'The provided matrix is:\n$transform'
     );
-    _transforms.add(_transforms.isEmpty ? transform : (transform * _transforms.last as Matrix4));
+    _localTransforms.add(_MatrixTransformPart(transform));
+  }
+
+  /// Pushes a new translation offset that is to be applied to all future
+  /// [HitTestEntry]s added via [add] until it is removed via [popTransform].
+  ///
+  /// This method is only to be used by subclasses, which must provide
+  /// coordinate space specific public wrappers around this function for their
+  /// users (see [BoxHitTestResult.addWithPaintOffset] for such an example).
+  ///
+  /// The provided `offset` should describe how to transform [PointerEvent]s from
+  /// the coordinate space of the method caller to the coordinate space of its
+  /// children. Usually `offset` is the inverse of the offset of the child
+  /// relative to the parent.
+  ///
+  /// [HitTestable]s need to call this method indirectly through a convenience
+  /// method defined on a subclass before hit testing a child that does not
+  /// have the same origin as the parent. After hit testing the child,
+  /// [popTransform] has to be called to remove the child-specific `transform`.
+  ///
+  /// See also:
+  ///
+  ///  * [pushTransform], which is similar to [pushOffset] but allows general
+  ///    transform besides translation.
+  ///  * [BoxHitTestResult.addWithPaintOffset], which is a public wrapper
+  ///    around this function for hit testing on [RenderBox]s.
+  ///  * [SliverHitTestResult.addWithAxisOffset], which is a public wrapper
+  ///    around this function for hit testing on [RenderSliver]s.
+  @protected
+  void pushOffset(Offset offset) {
+    assert(offset != null);
+    _localTransforms.add(_OffsetTransformPart(offset));
   }
 
-  /// Removes the last transform added via [pushTransform].
+  /// Removes the last transform added via [pushTransform] or [pushOffset].
   ///
   /// This method is only to be used by subclasses, which must provide
   /// coordinate space specific public wrappers around this function for their
   /// users (see [BoxHitTestResult.addWithPaintTransform] for such an example).
   ///
   /// This method must be called after hit testing is done on a child that
-  /// required a call to [pushTransform].
+  /// required a call to [pushTransform] or [pushOffset].
   ///
   /// See also:
   ///
-  ///  * [pushTransform], which describes the use case of this function pair in
-  ///    more details.
+  ///  * [pushTransform] and [pushOffset], which describes the use case of this
+  ///    function pair in more details.
   @protected
   void popTransform() {
+    if (_localTransforms.isNotEmpty)
+      _localTransforms.removeLast();
+    else
+      _transforms.removeLast();
     assert(_transforms.isNotEmpty);
```

Comment:
This assert is a little odd. Why do we assert after removing that there's still something in _transforms? Shouldn't we assert before?

Reply:
The `_transforms` should never be empty, because there should at least be the pre-seeded identity matrix.

Of course it shouldn't be empty before the operation either, but that would be detected by `removeLast` anyway.